lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,414 @@
+#    define NAMESPACE_IMPL server
+#    define RxComm         RsComm
+#    define rxTicketAdmin  rsTicketAdmin
+
+struct RsComm;
+#else
+#    define NAMESPACE_IMPL client
+#    define RxComm         RcComm
+#    define rxTicketAdmin  rcTicketAdmin
+
+struct RcComm;
+#endif // IRODS_TICKET_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+
+/**
+ * @brief Namespace to contain the API for tickets
+ *
+ */
+namespace irods::administration::ticket
+{
+    inline struct admin_tag
+    {
+    } admin;
+
+    enum class ticket_type
+    {
+        read,
+        write
+    };
+
+    enum class ticket_property
+    {
+        uses_count,
+        write_file,
+        write_byte
+    };
+
+    void execute_ticket_operation(RxComm& conn,
+                                  std::string_view command,
+                                  std::string_view ticket_identifier,
+                                  std::string_view command_modifier1,
+                                  std::string_view command_modifier2,
+                                  std::string_view command_modifier3,
+                                  std::string_view command_modifier4,
+                                  bool run_as_admin);
+
+    /**
+     * @brief Create a ticket object
+     *
+     * @param conn Communication object
+     * @param _type Specify type for ticket
+     * @param obj_path Path for the object
+     * @param ticket_name Name of the ticket (if provided)
+     * @return int Error code
+     */
+    void create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path, std::string_view ticket_name);
+    std::string create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    void create_ticket(admin_tag,
+                       RxComm& conn,
+                       ticket_type _type,
+                       std::string_view obj_path,
+                       std::string_view ticket_name);
+    std::string create_ticket(admin_tag, RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    /**
+     * @brief Delete a ticket
+     *
+     * @param conn Communication Object
+     * @param ticket_name Name of the ticket
+     * or
+     * @param ticket_ID ID for the ticket that needs to be deleted
+     * @return int Error Code
+     */
+    void delete_ticket(RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(RxComm& conn, int ticket_ID);
+
+    void delete_ticket(admin_tag, RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(admin_tag, RxComm& conn, int ticket_ID);
+
+    struct users_constraint
+    {
+        const std::string_view user_name;
+    };
+
+    struct groups_constraint
+    {
+        const std::string_view group_name;
+    };
+
+    struct hosts_constraint
+    {
+        const std::string_view host_name;
+    };
+
+    struct ticket_property_constraint
+    {
+        const ticket_property desired_property;
+        int value_of_property = -1;
+    };
```

Comment:
Do you actually need this?

You could encode the property to modify into the type itself (e.g. `use_count_constraint`, `n_writes_to_data_object_constraint`, `n_writes_constraint`). All of these types would simply hold an integer.

Reply:
Just did this :)